### PR TITLE
Add LogarithmicMapper

### LINKED ISSUE #6: Logarithmic-Mapper (spin)
Migrated from Enterprise Github.
Creator: @mrossinek

### What is the expected behavior?

Implement and test the `LogarithmicMapper`.
This issue breaks-out of #18 and can be used to assign and discuss related topics.

- [ ] Implementation
- [ ] Unittest
- [ ] Documentation

### PR DESCRIPTION
**Pull request by a-matsuo**
### Summary
Implemented a LogarithmicMapper class for a new QubitMapper.
- [x] unittests
- [x] documentation

### Details and comments
resolve #6

### CODE REVIEW

**Comment by a-matsuo** [comment]
@pbark It looks working, but I'm not sure whether the result is correct. Can we have simple test cases like the following ones in `test_linear_mapper.py`?   I checked the existing unittests for random operators, but it's hard to re-implement those tests on the current Qiskit. 

```
    spin_op1 = SpinOp([("Y_0^2", -0.432 + 1.32j)], 0.5, 1)
    ref_qubit_op1 = (-0.054 + 0.165j) * (I ^ I) + (0.054 - 0.165j) * (Z ^ Z)

    spin_op2 = SpinOp([("X_0 Z_0 I_1", -1.139 + 0.083j)], 0.5, 2)
    ref_qubit_op2 = (0.010375 + 0.142375j) * (I ^ I ^ Y ^ X) + (-0.010375 - 0.142375j) * (
        I ^ I ^ X ^ Y
    )
```

File: .pylintdict
Comment (mrossinek) [inline .pylintdict:153]:
```suggestion
guglielmo
```
The name in the citation is misspelled.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:35]: Besides the misspelled name, I would also suggest to use the more up-to-date reference: https://journals.aps.org/prd/abstract/10.1103/PhysRevD.102.094501

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:44]: If this can only take a limited set of strings I suggest we extract these into an `Enum`

Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:44]:
```suggestion
                Must be one of ['upper', 'lower']. This parameter sets whether
```

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:101]: This return type is actually much stricter. Above, you are unpacking the result of this function into four variables. Inspecting the code below, it is indeed the case that this list is always of length 4. I suggest you make this evident by returning a tuple and hard-code the length of 4 into the documented return type.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:78]: I do not understand the need for the `reduce()` call. Can you please explain why this is necessary or update the code if it is not.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:86]: How about the following instead?
```suggestion
                if len(operator_on_spin_i) > 0:
```
The `> 0` is even optional in both cases.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:88]:
```suggestion
                    operatorlist.append(single_operator_on_spin_i)
```
Or why is the `.reduce()` strictly necessary? If it is optional in terms of mathematics, how does performance change with/without it?

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:142]: This method does not use `self` and instead get's passed `embed_padding` and `embed_location` as input arguments.
I suggest you either make this a `@staticmethod` or change the latter and remove those duplicate arguments in favor of accessing them via `self`.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:184]: This seems to me like a lot of copying will take place, however, I am not 100% certain on the internals of numpy.
That said, I believe there may exist an easier implementation using `np.block` (see the first example here: https://numpy.org/doc/stable/reference/generated/numpy.block.html)

File: test/mappers/second_quantization/test_logarithmic_mapper.py
Comment (mrossinek) [inline test/mappers/second_quantization/test_logarithmic_mapper.py:74]:
```suggestion
```

**Review by mrossinek — DISMISSED** [review]
Thank you @a-matsuo for the hard work on this! I have a few comments below. Please let me know here or offline if I need to clarify anything :+1:

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (ikkoham) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:126]: Is it possible to implement this without lambda functions? Since this is a method, it should be able to take parameters inside.

File: test/mappers/second_quantization/test_logarithmic_mapper.py
Comment (ikkoham) [inline test/mappers/second_quantization/test_logarithmic_mapper.py:73]: `print` is used only for debug?

**Review by ikkoham — COMMENTED** [review]
LGTM. Just minor comments.

**Comment by woodsp-ibm** [comment]
Don't forget the release note!

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (woodsp-ibm) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:35]: Could we add something in the class docstring that briefly describes the embedding in a larger matrix, I see something in the map function but when I look  at the constructor it talks about embed_padding and embed_location and there is no immediate text to put these into context here for the reader. And I wonder if simply `padding` and `location` would be easier names if the text describes these as referring to the embedded matrix.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:87]: Oh I overlooked the fact that the list * integer multiplication will chain the list. Yeh this makes sense now, thanks! :+1:

**Comment by a-matsuo** [comment]
@mrossinek @ikkoham @woodsp-ibm 
Thank you for all the comments! I fixed the codes and reflected all the comments. Could you please review the code again?

File: releasenotes/notes/logarithmic_mapper-f655fc66d319ba08.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/logarithmic_mapper-f655fc66d319ba08.yaml:3]:
```suggestion
    Adds :class:`qiskit_nature.mappers.second_quantization.LogarithmicMapper`.
```
This needs to refer to the class itself not the module that its in (the release note showed the term highlighted but it was not a link). I guess we want to show the full path here for the context where the class is, that's fine I think.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (woodsp-ibm) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:48]: I can't help but wonder, given that this only has two values, whether a boolean might be "better". It would not need to add an extra class/enum type that needs to be documented etc i.e.
```
embed_upper: bool = True
```
Where if False then it does lower.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (woodsp-ibm) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:224]: As a bool this error test would then not be necessary

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:48]: I like the idea of using a boolean for this better than my original Enum suggestion, too :+1: Thanks @woodsp-ibm!

**Review by mrossinek — COMMENTED** [review]
I agree with Steve's comments and other than that this LGTM :+1:

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (a-matsuo) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:48]: I agree with @woodsp-ibm. Somehow, I never thought about using a boolean for 2 choices.

**Comment by a-matsuo** [comment]
Thank you so much guys! The PR is ready again now.

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:35]:
```suggestion
    [1]: S. V. Mathis, G. Mazzola and I. Tavernelli (2020).
```
This makes the use of first- and surnames consistent. I am also still not sure about the `(2020)` at the end.. @pbark do you have a suggestion here?

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:51]: Please reformat this to a 100-width text block

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (mrossinek) [inline qiskit_nature/mappers/second_quantization/logarithmic_mapper.py:54]: This will likely need to be wrapped in some sort of block in order for Sphinx to display this properly. Same below

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (a-matsuo) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py]: Thank you so much for reviewing the PR! I fixed everything. 

I'm not sure about the `(2020)` either. I just copied from this page, https://journals.aps.org/prd/abstract/10.1103/PhysRevD.102.094501. If we have something more appropriate, I have no problem to change it.  @mrossinek @pbark

Comment (mrossinek) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py]: I thought about this some more but forgot to update my comment here. How about we leave the names on a separate line and append the `(2020).` at the end of last line which indices the journal, volume and pages.

Comment (mrossinek) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py]: I think rather this:
```
Phys. Rev. D, 102 (9), 094501 (2020).
https://doi.org/10.1103/PhysRevD.102.094501
```

**Review by mrossinek — COMMENTED** [review]
Other than this final formatting comment above, this LGTM :+1:

File: qiskit_nature/mappers/second_quantization/logarithmic_mapper.py
Comment (a-matsuo) [reply qiskit_nature/mappers/second_quantization/logarithmic_mapper.py]: Thx! I fixed it.

**Review by mrossinek — APPROVED** [review]
Let's await a review by @woodsp-ibm and/or @pbark before merging :+1:

**Review by ikkoham — APPROVED** [review]
LGTM

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**
